### ISSUE TITLE
Incorrect lag assumptions after channels

### ISSUE THREAD

**Issue by slappaww** [simulationcraft/simc#8829]
From previous erroneously closed issue: https://github.com/simulationcraft/simc/issues/8825

I received the response 

"When you are casting a spell or are on the GCD, the game lets you queue you next spell, which will automatically cast immediately after your cast or the GCD finishes. If you try to cast a spell during a channel, it will simply interrupt the channel early resulting in lost ticks from the channel. Because casting spells immediately following a channel with no delay is unrealistic, simc adds a random delay after a channel ends before the next action is scheduled to model a human casting their next spell after the channel ends (the timing will not be exactly the same every time). Past testing has shown that the default setting of an average of 250 milliseconds of delay for channels is quite reasonable for this game"

This is trivially incorrect. 

Some examples from my recent log:

RSK cast 1s (GCD) after FoF. Sims assume 1.3s making fists_of_fury_cancel strictly a loss as it has 300ms of dead time that doesn't exist in game
![image](https://github.com/simulationcraft/simc/assets/42203447/00909a09-4bc3-459d-8a9b-c38b1f05f5e7)

Very next cast of fof into stomp is .013s delay 
![image](https://github.com/simulationcraft/simc/assets/42203447/4edfbb95-5e17-4332-8bac-9374fef7f07b)

I can with 100% consistency get this within 30ms vs the ~300ms default in simc which has significant APL impacts

SCK 30ms slippage
![image](https://github.com/simulationcraft/simc/assets/42203447/edb5fea4-f92c-4890-9d31-8611c8239eb3)

SCK 0ms slippage
![image](https://github.com/simulationcraft/simc/assets/42203447/9434a179-99af-423f-a388-eeb126cc0f27)

SCK is somewhat unique in that ticks do not cancel even if another GCD is pressed (as long as it is a non channel)

Another log highlighting a repeated lack of meaningful slippage on GCD cancels of FoF as well as SCK
![image](https://github.com/simulationcraft/simc/assets/42203447/77173ec6-8dbc-4f06-bb2a-a472db482e8d)

The exception is that SCK->SCK is not queue-able, but FoF into SCK is

**Comment by Dorovon** [comment]
You are queuing spells after the GCD ends there, not at the end of a channel. As I said, the game lets you queue spells in that situation. Channel lag is used in simc when the APL wishes to cast a spell immediately after a tick or at the end of a channel. All actions in the APL have the `interrupt_immediate=1` option available, which will queue the action at the end of the GCD with no channel lag. The `interrupt_global=1` option may also be needed in these cases to ensure that all higher priority actions are checked.

By default without those options, a higher priority action will interrupt the channel after the next tick following the GCD plus the channel lag. The tradeoff here is that you get one extra tick of the channel, but end up channeling the ability for slightly longer (including the channel lag penalty).

I do not play Monk at all, but simply adding both of those options to all channels with `interrupt=1` in the APL results in lost DPS. You can add them to actions you're interested in yourself if you want to compare behavior with the two modes.

**Comment by slappaww** [comment]
This just doesn't work as you are stating. 

As an example, using interrupt_immediate=1 I still get the following 
![image](https://github.com/simulationcraft/simc/assets/42203447/ef3aa75a-bd05-4f6d-92fe-36f92a6c5eba)

100ms of non existent delay is being created

There is an initial tick at ~3.014, another tick at ~3.53 immediately followed by an interrupt `Dot fists_of_fury fades from Player 'Fluffy_Pillow'`

Yet the follow up GCD has a 100ms delay arbitrarily. 

This is also true for Spinning Crane Kick which shouldn't require any interrupt conditionals given abilities are castable during the channel

This makes interrupt immediate purely a dps loss as normal interrupt just gets the 3rd tick at ~4.1

**Comment by Dorovon** [comment]
You also need `interrupt_global=1` so that higher priority actions from all action lists can cause the interrupt. This screenshot is from the default T31 profile this `,interrupt_immediate=1,interrupt_global=1` attached the lines after `interrupt=1`.
![image](https://github.com/simulationcraft/simc/assets/50294688/30929c41-771d-4ef0-8afe-5c0a5e22e156)

**Comment by Dorovon** [comment]
As far as Spinning Crane Kick goes, that's a special case that I don't know anything about since I don't develop Monk.

**Comment by renanthera** [comment]
For abilities that can be cast during SCK, `gcd_lag` is expected, whereas for abilities that cannot be cast during SCK, `world_lag` is used.

**Comment by renanthera** [comment]
Tl;dr if you use the correct action options, you will get the behaviour you desire. If you do not use the correct action options, you will not get the behaviour you desire.

Peruse the following simple APL cases for all of the FoF combinations. As Dorovon noted, with multiple action lists `interrupt_global` may be required to get the desired behaviour.
```
# 1
monk.initial_chi=5
world_lag=0
world_lag_stddev=0
override.bloodlust=0

actions+=/tiger_palm,if=time>0.1
actions+=/fists_of_fury

# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.874 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 1.748 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 2.622 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 3.496 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 3.743 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action tiger_palm time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
```
```
# 2
monk.initial_chi=5
world_lag=0
world_lag_stddev=0
override.bloodlust=0

actions+=/tiger_palm,if=time>0.1
actions+=/fists_of_fury,interrupt=1

# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.874 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 1.748 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 1.924 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action tiger_palm time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
```

```
# 3
monk.initial_chi=5
world_lag=0
world_lag_stddev=0
override.bloodlust=0

actions+=/tiger_palm,if=time>0.1
actions+=/fists_of_fury,interrupt=1,interrupt_immediate=1

# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.874 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 1.005 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action tiger_palm time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
```
```
# 4
monk.initial_chi=5
world_lag=0
world_lag_stddev=0
override.bloodlust=0

actions+=/tiger_palm,if=time>0.1
actions+=/fists_of_fury_cancel

# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_cancel time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.874 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 1.300 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action tiger_palm time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
```
```
# 5
monk.initial_chi=5
world_lag=0
world_lag_stddev=0
override.bloodlust=0

actions+=/tiger_palm,if=time>0.1
actions+=/fists_of_fury_cancel,interrupt=1

# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_cancel time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.874 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 1.238 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action tiger_palm time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
```
```
# 6
monk.initial_chi=5
world_lag=0
world_lag_stddev=0
override.bloodlust=0

actions+=/tiger_palm,if=time>0.1
actions+=/fists_of_fury_cancel,interrupt=1,interrupt_immediate=1

# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_cancel time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.000 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 0.874 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action fists_of_fury_tick time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
# 1.216 New Action Execute Event: Player 'DF4_Monk_Windwalker' Action tiger_palm time_to_execute=0.000 (target=Fluffy_Pillow, marker=0)
```

Cases 1, 2 and 3 all exhibit radically different behaviour:
1) Completes full FoF channel. Delay until next cast ~250ms.
2) Completes ticks until the first tick occurs after GCD elapsed. Delay until next cast ~250ms.
3) Completes ticks until the GCD elapsed. Next cast is immediately queued.

Cases 4, 5 and 6 all exhibit a combination of behaviours found in cases 2 and 3:
- Ticks occur until the GCD elapsed like case 3.
- Delay until next cast ~250ms.

**Comment by slappaww** [comment]
So what behavior uses GCD lag? Option 3 is the desired outcome but uses world lag when that does NOT mimic in game behavior, which is sort of the entire point

Ex: with 100ms world lag
![image](https://github.com/simulationcraft/simc/assets/42203447/ff283488-a4bb-46ba-81bf-3d6546de4fb4)

This is NOT how it should work. The delay between FoF->TP should be equivalent to the delay of TP->RSK

Similarly testing with Spining Crane Kick:

![image](https://github.com/simulationcraft/simc/assets/42203447/ee9f3160-0b72-49b5-937d-dc0dc9430d93)

There should be no world lag impact for TP->SCK yet there inexplicably is. 

Preferably it would not require world lag overrides to properly sim Windwalker and presumably whatever other spec has incorrectly implemented channels.
